File: src/zglab_rag/api/contracts.py
"""Public API contracts for Phase 9A.

This module defines the narrow public request/response models for the
public API. The design principle is to expose only what is safe and
necessary for the visitor; internal diagnostics, scores, and paths
are never leaked.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, Field


class PublicStatus(StrEnum):
    """Public response status values."""

    ANSWERED = "answered"
    INSUFFICIENT_EVIDENCE = "insufficient_evidence"


class PublicErrorCode(StrEnum):
    """Public error codes for the error envelope."""

    INVALID_REQUEST = "INVALID_REQUEST"
    RATE_LIMITED = "RATE_LIMITED"
    SERVICE_BUSY = "SERVICE_BUSY"
    GENERATION_TIMEOUT = "GENERATION_TIMEOUT"
    PROVIDER_UNAVAILABLE = "PROVIDER_UNAVAILABLE"
    INTERNAL_ERROR = "INTERNAL_ERROR"
    # Phase 11 security error codes. Login failures deliberately never
    # distinguish unknown user / wrong password / disabled account.
    AUTHENTICATION_REQUIRED = "AUTHENTICATION_REQUIRED"
    INVALID_CREDENTIALS = "INVALID_CREDENTIALS"
    ACCOUNT_UNAVAILABLE = "ACCOUNT_UNAVAILABLE"
    CSRF_REJECTED = "CSRF_REJECTED"
    QUOTA_EXCEEDED = "QUOTA_EXCEEDED"
    SERVICE_DISABLED = "SERVICE_DISABLED"
    API_RETIRED = "API_RETIRED"
    # Phase 12D capability-policy codes: the web kill switch refused the
    # request, or the account lacks the web-research permission.
    CAPABILITY_DISABLED = "CAPABILITY_DISABLED"
    CAPABILITY_DENIED = "CAPABILITY_DENIED"


class PublicAskRequest(BaseModel):
    """Public ask request body.

    The request stays narrow: question plus the additive Phase 12D ``mode``
    selector (auto/personal/web). Arbitrary capability ids are never
    accepted; unknown fields are still rejected outright. Clients that only
    send ``question`` keep working unchanged (mode defaults to auto).
    """

    model_config = {"extra": "forbid"}

    question: str = Field(min_length=1)
    mode: Literal["auto", "personal", "web"] = "auto"


class PublicSource(BaseModel):
    """A citation source safe for public display.

    Internal fields like chunk_id, document_id, revision, scores, and
    absolute paths are never exposed.

    Phase 12D additive fields: ``origin`` distinguishes personal knowledge
    from web evidence; web sources carry the provenance-validated ``url``
    and ``domain`` (never model output). Personal sources keep the frozen
    Phase 9 shape with origin=personal and url/domain absent.
    """

    id: str  # Evidence ID (E1, E2, ...)
    title: str
    section: list[str]
    source_path: str
    origin: Literal["personal", "web"] = "personal"
    url: str | None = None
    domain: str | None = None


class PublicAskResponse(BaseModel):
    """Public ask response for answered or insufficient_evidence status.

    The response is narrow and stable. It never includes internal
    diagnostics, scores, provider details, or absolute paths.
    """

    request_id: str
    status: Literal["answered", "insufficient_evidence"]
    answer: str
    sources: list[PublicSource]


class PublicErrorDetail(BaseModel):
    """Error detail for the public error envelope."""

    code: PublicErrorCode
    message: str


class PublicErrorResponse(BaseModel):
    """Public error response envelope.

    All errors (validation, rate limit, timeout, internal) use this
    envelope. The response never includes traceback, exception class,
    or internal paths.
    """

    request_id: str
    error: PublicErrorDetail


# ---------------------------------------------------------------------------
# Phase 9B: SSE stream contracts
# ---------------------------------------------------------------------------


class PublicStreamStage(StrEnum):
    """Public lifecycle stages emitted as SSE status events.

    Only `completed` may carry the final validated answer; stage events
    carry nothing but request_id and the stage name.
    """

    ACCEPTED = "accepted"
    RETRIEVING = "retrieving"
    RESEARCHING = "researching"
    GENERATING = "generating"
    VALIDATING = "validating"
    COMPLETED = "completed"


class PublicStreamStatus(BaseModel):
    """Narrow SSE status event payload (accepted/retrieving/generating/validating).

    Intentionally contains no evidence content, scores, provider details,
    token usage or diagnostics.
    """

    request_id: str
    stage: PublicStreamStage


# completed reuses the Phase 9A public response; SSE error reuses the
# Phase 9A public error envelope. Aliases keep the stream contract explicit
# without duplicating schema definitions.
PublicStreamCompleted = PublicAskResponse
PublicStreamError = PublicErrorResponse


# ---------------------------------------------------------------------------
# Phase 11: authenticated API v2 contracts
# ---------------------------------------------------------------------------


class AuthLoginRequest(BaseModel):
    """Login request body; narrow and extra-field rejecting."""

    model_config = {"extra": "forbid"}

    username: str = Field(min_length=1, max_length=64)
    password: str = Field(min_length=1, max_length=256)


class AuthActivateRequest(BaseModel):
    """Activation / password-reset consumption body (single-use token)."""

    model_config = {"extra": "forbid"}

    token: str = Field(min_length=1, max_length=256)
    password: str = Field(min_length=1, max_length=256)


class AuthChangePasswordRequest(BaseModel):
    model_config = {"extra": "forbid"}

    current_password: str = Field(min_length=1, max_length=256)
    new_password: str = Field(min_length=1, max_length=256)


class AuthUserPublic(BaseModel):
    """The minimal user representation safe for the SPA."""

    username: str
    role: Literal["ADMIN", "USER"]


class AuthSessionResponse(BaseModel):
    """Successful login / me response.

    The CSRF token is session-bound and held by the SPA in memory only;
    the session token itself never appears in any response body.
    """

    request_id: str
    user: AuthUserPublic
    csrf_token: str


class AuthResultResponse(BaseModel):
    """Generic success envelope for logout / activate / change-password."""

    request_id: str
    result: Literal[
        "logged_out",
        "account_activated",
        "password_updated",
        "password_changed",
    ]

File: web/src/api/contracts.ts
/**
 * Public API contracts, mirroring the narrow Phase 9A/9B backend schemas.
 * Internal diagnostics, scores, provider details and token usage are never
 * part of these types and must never be rendered.
 */

export type PublicStatus = "answered" | "insufficient_evidence";

export type PublicErrorCode =
  | "INVALID_REQUEST"
  | "RATE_LIMITED"
  | "SERVICE_BUSY"
  | "GENERATION_TIMEOUT"
  | "PROVIDER_UNAVAILABLE"
  | "INTERNAL_ERROR"
  // Phase 11 security error codes.
  | "AUTHENTICATION_REQUIRED"
  | "INVALID_CREDENTIALS"
  | "ACCOUNT_UNAVAILABLE"
  | "CSRF_REJECTED"
  | "QUOTA_EXCEEDED"
  | "SERVICE_DISABLED"
  | "API_RETIRED";

/** Frontend-only code for fetch/network/protocol failures. */
export type FrontendErrorCode = PublicErrorCode | "NETWORK";

export interface PublicSource {
  id: string;
  title: string;
  section: string[];
  source_path: string;
}

export interface PublicAskResponse {
  request_id: string;
  status: PublicStatus;
  answer: string;
  sources: PublicSource[];
}

export interface PublicErrorDetail {
  code: PublicErrorCode;
  message: string;
}

export interface PublicErrorResponse {
  request_id: string;
  error: PublicErrorDetail;
}

export type StreamStage = "accepted" | "retrieving" | "generating" | "validating";

export interface PublicStreamStatus {
  request_id: string;
  stage: StreamStage;
}

/** Typed union of everything the SSE stream can deliver. */
export type StreamEvent =
  | { kind: "stage"; stage: StreamStage; requestId: string }
  | { kind: "completed"; completed: PublicAskResponse }
  | { kind: "error"; error: PublicErrorResponse };

export const QUESTION_MAX_LENGTH = 1000;

/** Visitor-facing stage labels. No internal implementation details. */
export const STAGE_LABELS: Record<StreamStage, string> = {
  accepted: "已接收问题…",
  retrieving: "正在检索公开知识库…",
  generating: "正在整理回答…",
  validating: "正在核验引用…",
};

/** Visitor-facing error messages; raw server messages are never shown. */
export const ERROR_LABELS: Record<FrontendErrorCode, string> = {
  INVALID_REQUEST: "请求内容似乎有问题，请检查后重试。",
  RATE_LIMITED: "请求有点频繁，请稍后再试。",
  SERVICE_BUSY: "当前正在处理其他请求，请稍后再试。",
  GENERATION_TIMEOUT: "本次回答耗时过长，可以稍后重试。",
  PROVIDER_UNAVAILABLE: "回答服务暂时不可用，请稍后再试。",
  INTERNAL_ERROR: "服务暂时出现问题，请稍后再试。",
  AUTHENTICATION_REQUIRED: "请先登录后再使用助手。",
  INVALID_CREDENTIALS: "用户名或密码错误。",
  ACCOUNT_UNAVAILABLE: "账号当前不可用，请联系管理员。",
  CSRF_REJECTED: "请求未通过安全校验，请刷新页面后重试。",
  QUOTA_EXCEEDED: "今日或本分钟的使用额度已达上限，请稍后再试。",
  SERVICE_DISABLED: "回答服务当前已被临时关闭，请稍后再试。",
  API_RETIRED: "该接口版本已停用，请刷新页面。",
  NETWORK: "网络连接异常，请稍后重试。",
};

export function isPublicErrorCode(value: unknown): value is PublicErrorCode {
  return (
    value === "INVALID_REQUEST" ||
    value === "RATE_LIMITED" ||
    value === "SERVICE_BUSY" ||
    value === "GENERATION_TIMEOUT" ||
    value === "PROVIDER_UNAVAILABLE" ||
    value === "INTERNAL_ERROR" ||
    value === "AUTHENTICATION_REQUIRED" ||
    value === "INVALID_CREDENTIALS" ||
    value === "ACCOUNT_UNAVAILABLE" ||
    value === "CSRF_REJECTED" ||
    value === "QUOTA_EXCEEDED" ||
    value === "SERVICE_DISABLED" ||
    value === "API_RETIRED"
  );
}

export function isStreamStage(value: unknown): value is StreamStage {
  return (
    value === "accepted" ||
    value === "retrieving" ||
    value === "generating" ||
    value === "validating"
  );
}
